asf_public_discourse_home_decarbonisation/analysis/topic_analysis/topic_analysis_heat_pump_keywords.py

```diff
@@ -0,0 +1,227 @@
+- to the titles of the threads where heat pumps are mentioned (either in the title, the text or in replies);
+- We fix a random seed to ensure reproducibility of the results and use;
+- We set specific BERTopic parameters after having evaluated the model with different configurations.
+
+After running the topic model, we then visualise the topics and the documents associated with them, and we analyse the most common topics.
+
+To transform this script into a jupyter notebook, we can use the following commands:
+- `jupytext --to notebook topic_analysis_heat_pump_keywords.py`
+- If the correct kernel does not come up (`asf_public_discourse_home_decarbonisation`), please run the following in your terminal:  `python -m ipykernel install --user --name=asf_public_discourse_home_decarbonisation`
+"""
+
+# %%
+from bertopic import BERTopic
+from umap import UMAP
+from bertopic.representation import KeyBERTInspired
+from sklearn.feature_extraction.text import CountVectorizer
+import pandas as pd
+import matplotlib.pyplot as plt
+from asf_public_discourse_home_decarbonisation.getters.mse_getters import get_mse_data
+from asf_public_discourse_home_decarbonisation.getters.bh_getters import get_bh_data
+from asf_public_discourse_home_decarbonisation.config.keywords_dictionary import (
+    keyword_dictionary,
+)
+from asf_public_discourse_home_decarbonisation.utils.topic_analysis_utils import (
+    create_bar_plot_most_common_topics,
+    get_outputs_from_topic_model,
+)
+
+# %% [markdown]
+# ## Loading & filtering data
+
+# %%
+mse_data = get_mse_data(
+    category="all", collection_date="2023_11_15", processing_level="raw"
+)
+
+# %%
+bh_data = get_bh_data(category="all", collection_date="24_02_01")
+
+# %%
+# Replacing abbreviations
+mse_data["title"] = mse_data["title"].apply(
+    lambda x: x.lower()
+    .replace("ashps", "ashp")
+    .replace("ashp", "air source heat pump")
+    .replace("gshps", "gshp")
+    .replace("gshp", "ground source heat pump")
+    .replace("hps", "hp")
+    .replace("hp", "heat pump")
+    .replace("ufh", "under floor heating")
+)
+mse_data["text"] = mse_data["text"].apply(
+    lambda x: x.lower()
+    .replace("ashps", "ashp")
+    .replace("ashp", "air source heat pump")
+    .replace("gshps", "gshp")
+    .replace("gshp", "ground source heat pump")
+    .replace("hps", "hp")
+    .replace("hp", "heat pump")
+    .replace("ufh", "under floor heating")
+)
+# Replacing abbreviations
+bh_data["title"] = (
+    bh_data["title"]
+    .astype(str)
+    .apply(
+        lambda x: x.lower()
+        .replace("ashps", "ashp")
+        .replace("ashp", "air source heat pump")
+        .replace("gshps", "gshp")
+        .replace("gshp", "ground source heat pump")
+        .replace("hps", "hp")
+        .replace("hp", "heat pump")
+        .replace("ufh", "under floor heating")
+    )
+)
+bh_data["text"] = (
+    bh_data["text"]
+    .astype(str)
+    .apply(
+        lambda x: x.lower()
+        .replace("ashps", "ashp")
```

Comment:
for ashps to ashp and gshps to gshp, why wouldn't we just replace ashps and gshps with "air source heat pumps" and "ground source heat pumps" respectively?

Reply:
just made it this way so we always have the singular form for all of these acronyms, but no big difference in the end output.

Reply:
It won't make a big difference in the end to the output, only difference will be with sentences like "Why are there so many ashps" will go to "why are there so many air source heat pump" instead of "why are there so many air source heat pumps".

Reply:
shouldn't be problematic for clustering titles, but i changed it now to:

```
mse_data["title"] = mse_data["title"].apply(
    lambda x: x.lower()
    .replace("ashps", "air source heat pumps")
    .replace("ashp", "air source heat pump")
    .replace("gshps", "ground source heat pumps")
    .replace("gshp", "ground source heat pump")
    .replace("hps", "heat pumps")
    .replace("hp", "heat pump")
    .replace("ufh", "under floor heating")
)
```